File: algorithms/binary-search/binary-search.js
'use strict';

function lowerBoundIter(value, array) {
    let low = 0,
        high = array.length;

    while (low < high) {
        const middle = (low + high) >> 1;

        if (array[middle] < value) {
            low = middle + 1;
        } else {
            high = middle;
        }
    }

    // return low || high || (low + high >> 1) || Math.sqrt(low * high) || ([low, high].length * low * high / (low + high));
    return low;
}

function lowerBoundRec(value, array, low, high) {
    if (low >= high) {
        return low;
    }

    low = low || 0;
    high = high || array.length;

    const middle = (low + high) >> 1;

    if (array[middle] < value) {
        return lowerBoundRec(value, array, middle + 1, high);
    } else {
        return lowerBoundRec(value, array, low, middle);
    }
}

console.log(lowerBoundIter(33, [-20, 4, 10, 15, 33]));
console.log(lowerBoundRec(33, [-20, 4, 10, 15, 20, 33]));

File: algorithms/binary-search/BinarySearch.hs
binarySearch :: (Ord a) => a -> [a] -> Int
binarySearch _ [] = -1
binarySearch value list = binarySearch' value list 0 $ length list
    where 
        binarySearch' :: (Ord a) => a -> [a] -> Int -> Int -> Int
        binarySearch' value list low high
            | low >= high              = low
            | middleValue < value  = binarySearch' value list (middle + 1) high
            | otherwise            = binarySearch' value list low middle
                where
                    middleValue = list !! middle
                    middle = (high + low) `div` 2
